Support allreduce in inductor

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90813
* #90804

1) use  'traceable_allreduce' op from prev PR, add inductor lowering for it
2) lowering constructs a pair of [AllReduce, Wait] kernels, and treats AllReduce as a hidden one (from user)  only consumed by 'Wait'.  User (who called `traceable_allreduce`) gets a 'Wait' kernel wrapped in a TensorBox, which refers to and has the layout of the original input to the AllReduce.

In effect, this makes AllReduce act like an inplace operation (no allocations happen), but decouples the start and the finish.

Next, I need to make Inductor's scheduler smart enough to prefer scheduling the Wait as late as possible.  For now, it could get scheduled anytime before first use. 

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

wconstab committed:
Support allreduce in inductor (hackily)

[ghstack-poisoned]

ngimel commented on torch/_inductor/ir.py:
```diff
@@ -3349,6 +3349,34 @@ def get_template_tiling(self):
+        x: "TensorBox",
+    ):
+        return AllReduce(
+            layout=MutationLayout(x),
```
You probably should realize `x` before calling AllReduce, otherwise you can't get its layout and stride

wconstab replied:
bingo!

wconstab committed:
Update on "Support allreduce in inductor (hackily)"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Support allreduce in inductor (hackily)"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

aazzolini commented on torch/_inductor/ir.py:
```diff
@@ -3349,6 +3349,46 @@ def get_template_tiling(self):
+class AllReduce(ExternKernelAlloc):
+    kernel = "torch.ops.c10d.traceable_allreduce"
+    #1 constant_args show up in codegen, use those if needed
+    #2 allreduce_ can't be called from python, without fixing bindings
```
what does this mean? can't you call dist.allreduce directly from python?

wconstab replied:
i switched to dist.allreduce - but dist.allreduce returns a work.

i wanted to initially try calling `c10d.allreduce_`  -- it should be possible, but the bindings are broken.

this isn't blocking me anymore; i think the `dist` route is going to work.

aazzolini commented on torch/_inductor/ir.py:
```diff
@@ -3349,6 +3349,46 @@ def get_template_tiling(self):
+    #1 constant_args show up in codegen, use those if needed
+    #2 allreduce_ can't be called from python, without fixing bindings
+    #3 try implementing traceable_allreduce and adding POD types to api for rank/size?
+    #4 think about what to do for cpu
```
is this still an issue if you're using c10d that deals with backend (nccl/gloo) selection?

wconstab replied:
i think this concern goes away if the `dist` route works fully (and is efficient)

aazzolini commented on torch/_inductor/ir.py:
```diff
@@ -3349,6 +3349,46 @@ def get_template_tiling(self):
+    #2 allreduce_ can't be called from python, without fixing bindings
+    #3 try implementing traceable_allreduce and adding POD types to api for rank/size?
+    #4 think about what to do for cpu
+    #  also how to handle comm stream
```
is this an issue if c10d is used?

aazzolini commented on torch/_inductor/ir.py:
```diff
@@ -3349,6 +3349,46 @@ def get_template_tiling(self):
+    #     # allreduce_ expects a list of tensor inputs...
+    #     args[0] = f"[{args[0]}]"
+    #     wrapper.writeline(
+    #         f"{self.get_name()}, _ = {self.kernel}({', '.join(args)})"
```
you probably don't want to call traceable_allreduce here but rather dist.allreduce directly?

wconstab committed:
Update on "Support allreduce in inductor (hackily)"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Support allreduce in inductor (hackily)"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Support allreduce in inductor (hackily)"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Support allreduce in inductor (hackily)"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Support allreduce in inductor (hackily)"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Support allreduce in inductor (hackily)"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wconstab commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -260,6 +260,11 @@ def __init__(self):
                 from torch import empty_strided, as_strided, device
                 from {codecache.__name__} import AsyncCompile
 
+                # multiprocess triton crashes without setting default device
```
hopefully this is fixed by https://github.com/pytorch/pytorch/pull/90934 and i can remove this

wconstab commented on torch/_inductor/graph.py:
```diff
@@ -188,6 +190,12 @@ def register_buffer(self, buffer: ir.ComputedBuffer):
         self.name_to_buffer[name] = buffer
         return name
 
+    def register_work(self, work):
```
i didn't end up using this after all; but i might still want a way to give special naming (if not handling) to work objects.  it's goofy to see code like 

```buf2 = dist.allreduce()
buf2.wait()
```

wconstab committed:
Update on "Support allreduce in inductor"


1) use  'traceable_allreduce' op from prev PR, add inductor lowering for it
2) lowering constructs a pair of [AllReduce, Wait] kernels, and treats AllReduce as a hidden one (from user)  only consumed by 'Wait'.  User (who called `traceable_allreduce`) gets a 'Wait' kernel wrapped in a TensorBox, which refers to and has the layout of the original input to the AllReduce.

In effect, this makes AllReduce act like an inplace operation (no allocations happen), but decouples the start and the finish.

Next, I need to make Inductor's scheduler smart enough to prefer scheduling the Wait as late as possible.  For now, it could get scheduled anytime before first use. 

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]
